zkevm-circuits/src/evm_circuit/execution/exp.rs

```diff
@@ -0,0 +1,304 @@
+        let (exponent_lo, exponent_hi) = (
+            from_bytes::expr(&exponent_rlc.cells[0x00..0x10]),
+            from_bytes::expr(&exponent_rlc.cells[0x10..0x20]),
+        );
+        // Extract low and high bytes of the exponentiation result.
+        let (exponentiation_lo, exponentiation_hi) = (
+            from_bytes::expr(&exponentiation_rlc.cells[0x00..0x10]),
+            from_bytes::expr(&exponentiation_rlc.cells[0x10..0x20]),
+        );
+
+        // We simplify constraints depending on whether or not the exponent is 0 or 1.
+        // In order to do this, we build some utility expressions.
+        let exponent_lo_is_zero = IsZeroGadget::construct(cb, exponent_lo.clone());
+        let exponent_hi_is_zero = IsZeroGadget::construct(cb, exponent_hi.clone());
+        let exponent_is_zero_expr =
+            and::expr([exponent_lo_is_zero.expr(), exponent_hi_is_zero.expr()]);
+        let exponent_lo_is_one = IsEqualGadget::construct(cb, exponent_lo.clone(), 1.expr());
+        let exponent_is_one_expr =
+            and::expr([exponent_lo_is_one.expr(), exponent_hi_is_zero.expr()]);
+
+        let zero_rlc = cb.query_word();
+        cb.require_zero(
+            "base * base + c == base^2 (c == 0)",
+            sum::expr(&zero_rlc.cells),
+        );
+        let base_sq = cb.query_word();
+
+        // If exponent == 0, base^exponent == 1, which implies:
+        // 1. Low bytes of exponentiation == 1
+        // 2. High bytes of exponentiation == 0
+        cb.condition(exponent_is_zero_expr.clone(), |cb| {
+            cb.require_equal(
+                "exponentiation == 1 if exponent == 0 (lo == 1)",
+                exponentiation_lo.clone(),
+                1.expr(),
+            );
+            cb.require_equal(
+                "exponentiation == 1 if exponent == 0 (hi == 0)",
+                exponentiation_hi.clone(),
+                0.expr(),
+            );
+        });
+        // If exponent == 1, base^exponent == base, which implies:
+        // 1. Low bytes of exponentiation == low bytes of base.
+        // 2. High bytes of exponentiation == high bytes of base.
+        cb.condition(exponent_is_one_expr.clone(), |cb| {
+            cb.require_equal(
+                "exponentiation == base if exponent == 1 (lo)",
+                exponentiation_lo.clone(),
+                base_lo.clone(),
+            );
+            cb.require_equal(
+                "exponentiation == base if exponent == 1 (hi)",
+                exponentiation_hi.clone(),
+                base_hi.clone(),
+            );
+        });
+        // If exponent > 1, i.e. exponent != 0 && exponent != 1:
+        // We do two lookups to the exponentiation table. If exponent == 2, there is
+        // only a single step in the exponentiation by squaring trace. In this
+        // case, is_first == is_last == true for that step.
+        let single_step = cb.query_cell();
+        cb.condition(
```

Comment:
yes, that's true. For compatibility of specs and circuits, that would be good. I'll make that change in the Specs repository.

Actually, now I realise that even if there is a single step (`exponent == 2`) then we're doing 2 lookups in the `EXP` gadget (`zkevm-circuits`), both lookups to the exact same row:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/53ef4605fec256f18f89d0f7ee78495daf5acffd/zkevm-circuits/src/evm_circuit/execution/exp.rs#L149-L165

While in the `zkevm-specs` repository, the condition on `exponent == 2` means we're doing 1 lookup for `exponent == 2` and 2 lookups for `exponent > 2`.

The question is:
1. Would it be more efficient to define an `IsEqual` gadget to compare `exponent` and `2`, and then conditionally do 1 or 2 lookups?
2. Or would it be more efficient to use a single cell (`single_step`) and always do 2 lookups (both being the same row in case `exponent == 2`)?

Reply:
What about this issue @adria0 ?